clis/teliod/src/comms.rs

```diff
@@ -0,0 +1,120 @@
+//! Code for handling inter process communication and abstracting away it's dependencies.
+
+use std::{fs, io::Write, path::Path};
+
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
+
+use anyhow::Result;
+use interprocess::{
+    local_socket::{
+        tokio::{Listener as LocalSocketListener, Stream as LocalSocketStream},
+        traits::tokio::{Listener, Stream},
+        ListenerOptions, ToFsName,
+    },
+    os::unix::local_socket::FilesystemUdSocket,
+};
+
+/// Struct for handling connections of the daemon's side of the IPC communication with the API.
+pub struct DaemonSocket {
+    /// The inner socket over which the actual communication is happening.
+    socket: LocalSocketListener,
+}
+
+impl DaemonSocket {
+    /// Binds the IPC socket to the specified address and returns a handle to the struct
+    /// containing the IPC socket.
+    ///
+    /// # Arguments
+    ///
+    /// * `ipc_socket_path` - Path of the IPC socket.
+    ///
+    /// # Returns
+    ///
+    /// An instance of Self wrapped inside of a Result.
+    pub fn new(ipc_socket_path: &Path) -> Result<Self> {
+        // Delete the socket file if it already exists
+        let _ = fs::remove_file(ipc_socket_path);
+        let socket = ListenerOptions::new()
+            .name(ipc_socket_path.to_fs_name::<FilesystemUdSocket>()?)
+            .create_tokio()?;
+
+        Ok(Self { socket })
+    }
+
+    /// Opens a connection on the IPC socket for listening to commands (requests) from the API and
+    /// returns them to be processed by the event loop.
+    /// The same connection is used to send responses back to the API when the event loop produces them.
+    /// Note that since a single socket is used, care should be taken to not deadlock it.
+    ///
+    /// # Returns
+    ///
+    /// A struct for receiving commands from the new IPC connection stream and responding to them.
+    pub async fn accept(&self) -> Result<DaemonConnection> {
+        let stream = self.socket.accept().await?;
+
+        Ok(DaemonConnection { stream })
+    }
+
+    /// Function for sending commands to the server.
+    /// Also retrieves the response from the IPC server, but blocks while waiting for it.
+    /// Note that endlines are used to indicate the end of a message so the server knows when to stop reading and
+    /// send a response, otherwise the server would wait until the API closes the socket and won't be able to send a response.
+    /// This is also sort of a deadlock risk, so should be handled with care.
+    ///
+    /// # Arguments
+    ///
+    /// * `addr` - Address (more specifically path, but left address for clarity) of the IPC socket.
+    /// * `cmd` - Command to be sent to the daemon.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing a response string.
+    pub async fn send_command(addr: &Path, cmd: &str) -> Result<String> {
```

Comment:
To some extent we depend on `cmd` not having any newlines in itself. Maybe it is worth veryfing it and enforcing this, e.g. by returning an error if there is a newline inside of cmd?

Reply:
But here? Only JSONed commands are sent here, I don't think they can contain any newlines. But I can add some check, sure ;)